feat: save assessment questions

joyce-shi opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Save Assessment Questions section on Create Assessment page](https://www.notion.so/uwblueprintexecs/Save-Assessment-Questions-section-on-Create-Assessment-page-a7843e084ec746ed9cfc2407b05f0e67?pvs=4)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Functionality for saving assessment questions

Note: I bumped the Typescript version to resolve an issue I was having defining optional array fields in interfaces, similar to the one described [here](https://github.com/microsoft/TypeScript/issues/41651).

<!-- What should the reviewer do to verify your changes? Describe expected results and include screenshots when appropriate -->
## Steps to test
1. no breaking changes
2. elements of the question editor save properly

joyce-shi committed:
add question frontend

joyce-shi committed:
question card fixes

joyce-shi committed:
lint

joyce-shi committed:
render question cards based on data

joyce-shi committed:
restructuring create assessmnet hierarchy

joyce-shi committed:
render question editor data dynamically

joyce-shi committed:
refactoring

joyce-shi committed:
some error validation

joyce-shi committed:
add button

joyce-shi committed:
validation logic

joyce-shi committed:
add empty editor error

joyce-shi committed:
Add response element type as subset of question element type

joyce-shi committed:
remove unused imports

joyce-shi committed:
fixes

joyce-shi committed:
delete unused file

joyce-shi committed:
addressing pr comments

joyce-shi committed:
minor fixes

joyce-shi committed:
checkpoint

joyce-shi committed:
changing question and short answer types to match backend

joyce-shi committed:
saving text and multiple choice question components

joyce-shi committed:
cleanup

joyce-shi committed:
temporary type issue fix

joyce-shi committed:
move types to api clients folder

joyce-shi committed:
refactor logic to question utils

joyce-shi committed:
refactor

joyce-shi committed:
bump ts version to fix interface bug

joyce-shi committed:
check for at least one question before saving

joyce-shi commented on frontend/src/APIClients/mutations/TestMutations.ts:
```diff
@@ -30,7 +30,7 @@ export const UNARCHIVE_TEST = gql`
+export const CREATE_NEW_ASSESSMENT = gql`
   mutation Create_Test($test: TestRequestDTO!) {
     createTest(test: $test) {
       id
```
updated the naming, since this is the same mutation we should use when pressing the **Publish** button on the Create Assessment page

joyce-shi commented on frontend/src/components/assessments/assessment-creation/BasicInformation.tsx:
```diff
@@ -17,70 +19,90 @@ import {
+                    gradeOptions.find(
+                      (option) => option.value === watch("grade"),
+                    ) || undefined
+                  }
```
I think we should probably split some of these `Select` components into separate components. I think there's a couple bugs related to dropdowns and also a ticket for refactoring some stuff into common components, so I think I'll work on all of these things together in a subsequent PR! So leaving in this component for now

joyce-shi commented on frontend/src/types/QuestionTypes.ts:
```diff
@@ -1,10 +1,41 @@
+  MULTIPLE_CHOICE = "MULTIPLE_CHOICE",
+  SHORT_ANSWER = "SHORT_ANSWER",
+  MULTI_SELECT = "MULTI_SELECT",
+}
```
Updated the naming of this values of this enum to match the corresponding backend enum. Then, in all the places where the value was used (i.e as a text label), used our existing formatting utils.

joyce-shi commented on frontend/src/types/QuestionTypes.ts:
```diff
@@ -1,10 +1,41 @@
+}
+
+export type QuestionElementDataType =
+  // remove string when all the elements have been implemented
```
There's some placeholder values for other elements right now, so leaving this in. Later, when we have all the elements implemented, can remove this line!

joyce-shi commented on frontend/src/utils/QuestionUtils.ts:
```diff
@@ -98,7 +106,52 @@ export const getQuestionTexts = (question: QuestionElement[]): string[] => {
+    options: data.options.map((option) => option.value),
+    answerIndex: data.options.findIndex((option) => option.isCorrect),
+  };
+};
```
for other question components, i just stored their `data` in their metadata types. but for multiple choice, since I needed an id field in its implementation, i just defined this extra helper function to convert it to its metadata type!

joyce-shi commented:
cc @cindy-531

carissa-tang commented on package-lock.json:
```diff
@@ -0,0 +1,15 @@
+{
```
is this change needed ? since we are updating frontend/yarn.lock already

joyce-shi replied:
removed!

carissa-tang commented on frontend/src/components/question-creation/question-elements/modals/multiple-choice/MultipleChoiceModal.tsx:
```diff
@@ -33,7 +33,7 @@ const MultipleChoiceModal = ({
 
   useEffect(() => {
-    setOptionCount(data ? data.optionCount : 0);
+    setOptionCount(data ? data.options.length : 0);
```
wait do we still need `optionCount` or can we just use `options.length`

joyce-shi replied:
i think it still makes sense to keep `optionCount`? bc we use it as a local variable to add elements to `options` (so if we just used `options.length` it would always be 0)

carissa-tang replied:
ah kk

carissa-tang reviewed (commented):
1. should we add a placeholder here
<img width="430" alt="image" src="https://user-images.githubusercontent.com/69697180/229665193-e6401774-efac-416e-ac96-db5c97ff4036.png">
2. can we clear these errors on change
<img width="868" alt="image" src="https://user-images.githubusercontent.com/69697180/229666545-392257ce-9543-4cf6-ad79-2568e83b6098.png">

joyce-shi committed:
Delete package-lock.json

joyce-shi committed:
addressing pr comments

joyce-shi committed:
add handling for multi select

joyce-shi commented:
addressed pr comments & added handling for multi-select (since jenna was able to get that merged yesterday!!) also removed `package-lock.json` since i think that was accidentally added

carissa-tang reviewed (approved):
LGTM!!
